torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,325 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
+  }
+
+  // returns true if either id is not a broadcast or
+  // the traversal has found a concretized axis for id
+  bool canConcretize(IterDomain* id) {
+    return !id->isBroadcast() || bcast_domain_map_.count(id);
+  }
+
+  // returns the concretized id recorded from traversal
+  IterDomain* concretized(IterDomain* id) {
+    TORCH_INTERNAL_ASSERT(canConcretize(id));
+    if (!id->isBroadcast())
+      return id;
+    return bcast_domain_map_.at(id);
+  }
+
+ private:
+  using MapType = std::unordered_map<IterDomain*, IterDomain*>;
+  MapType bcast_domain_map_;
+
+  // utility to inspect a pointwise operator and
+  // record concretize opportunities
+  void concretizePwOp(Expr* e);
+
+  // utility to record new concretize opportunity
+  void concretizeTo(IterDomain* id, IterDomain* To) {
+    TORCH_INTERNAL_ASSERT(id->isBroadcast() && !To->isBroadcast());
+    bcast_domain_map_[id] = concretized(To);
+  }
+
+  void handle(ReductionOp* rop) override {
+    concretizePwOp(rop->asExpr());
+  }
+
+  void handle(UnaryOp* uop) override {
+    concretizePwOp(uop->asExpr());
+  }
+
+  void handle(BinaryOp* bop) override {
+    concretizePwOp(bop->asExpr());
+  }
+
+  void handle(TernaryOp* top) override {
+    concretizePwOp(top->asExpr());
+  };
+};
+
+void ConcretizeDomain::concretizePwOp(Expr* e) {
+  TensorView* tv = *ir_utils::filterByType<TensorView>(e->outputs()).begin();
+
+  // won't use this function on a reduction op, so root domain only
+  std::vector<IterDomain*> io = tv->getRootDomain();
+
+  for (auto* i : ir_utils::filterByType<TensorView>(e->inputs())) {
+    std::vector<IterDomain*> ii =
+        TensorDomain::noReductions(i->getMaybeRFactorDomain());
+    TORCH_INTERNAL_ASSERT(ii.size() == io.size());
+
+    for (size_t it = 0; it < ii.size(); it++) {
+      if (!canConcretize(io[it]))
+        continue;
+
+      if (!canConcretize(ii[it]))
+        concretizeTo(ii[it], concretized(io[it]));
+    }
+  }
+} // ConcretizeDomain::pointWiseOpMap
+
+//! Models equality provable by the graph
+//!
+//! This traversal processes root domains only,
+//! equalities , e.g. :
+//!    T2 [i0,i1] = T1[i2,i3] + T0[i4,i5]
+//! will prove that i2 and i4 are equal in the sense that
+//!    i2.start = i4.start, i2.extent = i4.extent
+//! Depends on ConcretizeDomain, and equalities involving
+//! broadcast domains are defined based on the concretized version
+class ProveValEqual : public IterVisitor {
+ public:
+  ProveValEqual(Fusion* fusion) : cd_(fusion) {
+    traverseFrom(fusion, fusion->outputs(), false);
+  }
+
+  //! checks if two scalars are equal
+  //!
+  //! first checks if ScalarCheck has them equal,
+  //! next try to prove them equal from
+  //! the graph_traversal result
+  //!
+  //! \param a a symbolic value
+  //! \param b another value from the same fusion
+  //! \returns boolean representing if they are proven to be
+  //!          equal based on scalar check and graph traversal
+  bool areEqual(Val* a, Val* b) {
```

Comment:
`areEquivalent` ?

Reply:
Thanks for the suggestions. I think in the case of symbolic integers "equal" makes sense, but in the case of iterdomains equivalent is more accurate. I updated the functions to equivalent for iterdomains.